[JSC][Wasm][Debugger] Implement multi-VM stop-the-world debugging

hyjorc1 opened the pull request:
#### e51159c6bbedb25e0e9eeb217917dce4eb7bf327
<pre>
[JSC][Wasm][Debugger] Implement multi-VM stop-the-world debugging
<a href="https://bugs.webkit.org/show_bug.cgi?id=302699">https://bugs.webkit.org/show_bug.cgi?id=302699</a>
<a href="https://rdar.apple.com/164945623">rdar://164945623</a>

Reviewed by Keith Miller.

Implements comprehensive multi-VM stop-the-world Wasm debugger that stops ALL VMs
in the process when any VM hits a breakpoint or receives an interrupt.

Architecture:

* VMManager world modes coordinate multi-VM execution:
    - Running: Normal execution (all VMs running).
    - Stopped: All VMs paused at safe points for inspection.
    - RunOne: Single-step mode (target VM runs, others stopped).

* Two-phase callback system:
    - wasmDebuggerOnStop(): VMManager calls with ALL VMs stopped; returns resume mode.
    - wasmDebuggerOnResume(): VMManager calls after ALL VMs resume; releases barriers.

* Per-VM DebugState tracks execution state:
    - Running vs Stopped with stop reason (Prologue/Breakpoint/SystemCall).
    - StopData snapshot: PC, locals, stack, call frame for inspection.
    - Step-into event flags for cross-function stepping (call/throw).

* ExecutionHandler bridges LLDB commands to VM execution:
    - Detailed design in Debugger-Mutator-Protocol.md.
    - Two condition variables orchestrate debugger-mutator communication.
    - State machine (Replied/InterruptRequested/ContinueRequested/StepRequested/SwitchRequested).
    - Resume barrier prevents overlapping stop-the-world operations.
    - Target VM selection: prefers VM at prologue, falls back to triggering VM.

* Step-into design minimizes runtime changes:
    - step() sets event flag instead of resolving callee/handler immediately
    - Runtime resolves target naturally during execution (call dispatch, exception unwinding)
    - Callback sets breakpoint using resolved target, avoiding duplicate resolution logic
    - Reuses existing runtime mechanisms rather than duplicating complex resolution in debugger

Testing:

* Unit tests: ExecutionHandlerTest validates coordination logic.
* Integration tests: multi-vm test cases for same/different modules and functions.

Note that the WebAssembly debugger currently supports ARM64 platforms only.

Canonical link: <a href="https://commits.webkit.org/305160@main">https://commits.webkit.org/305160@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/442f46edfb04c7829905391b93c2a5a2ce3f4bee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137622 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9983 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48910 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/145380 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90595 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0cbb05b4-618d-4dcc-a747-378a438118ef) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10686 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10113 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105259 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140567 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7939 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123337 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86113 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1f5b781d-04e6-44b9-a6be-c016a778d281) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7568 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5291 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5963 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/129583 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116955 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41497 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148147 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/136143 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9666 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42048 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113641 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9683 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8146 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113984 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28946 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7497 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119577 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/64329 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9714 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37626 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168892 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9445 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73279 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/44060 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9654 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9506 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/JavaScriptCore/runtime/VM.cpp:
```diff
@@ -1885,4 +1886,15 @@ void VM::DrainMicrotaskDelayScope::decrement()
+Wasm::DebugState* VM::debugState()
+{
+    if (!m_debugState)
+        m_debugState = makeUnique<Wasm::DebugState>();
```
This might be a silly question, but since this is Wasm is it possible for multiple worker threads to access this at once?

hyjorc1 replied:
Excellent question! Thread safety here works in two phases:

1. Local VM Access (No Lock Required):
- Each VM's own mutator thread initializes and updates its DebugState when executing Wasm code (Examples: `setPrologueStopData`, `setBreakpointStopData`).
  - Always called by the VM's own thread.
- This is safe because each VM only touches its own DebugState.

2. Cross-VM Access (Lock Required):
- The debugger/debuggee thread reads/modifies ANY VM's DebugState under `ExecutionHandler::m_lock`.
- This happens during stop-the-world callbacks triggered by debuggee (debug target VM) when all other VMs are paused at safe points.
- While stopped, all non-debuggee VMs wait at VMManager::m_worldConditionVariable, so they can't race with debugger or target mutator access.

Key Safety Mechanism:
- The stop-the-world protocol ensures that when the debugger/debuggee thread accesses a DebugState (under lock), the owner VM is guaranteed to be stopped. This prevents races between:
  - Non-debuggee VM thread updating its own state.
  - Debugger/debuggee thread reading/modifying that state.

So yes, multiple threads can access it, but the lock + stop-the-world coordination ensures mutual exclusion.

Achierius commented on Source/JavaScriptCore/wasm/debugger/WasmQueryHandler.cpp:
```diff
@@ -209,17 +210,24 @@ bool QueryHandler::handleChunkedLibrariesResponse(size_t offset, size_t maxSize,
     VirtualAddress virtualReturnPC;
     unsigned frameIndex = 0;
 
+    // FIXME: Only supports consecutive wasm->wasm calls. Need to support interleaved wasm<->js calls (skipping stubs, including JS frames).
```
What will happen right now if you try to use this with interleaved frames? Will it assert with some message, just crash, or can it behave badly?

hyjorc1 replied:
So far, it only can walk through pure Wasm call chains: `Wasm->Wasm->Wasm->...`. And the walk stops at any non-Wasm frame.

Achierius commented on Source/JavaScriptCore/wasm/debugger/WasmExecutionHandler.cpp:
```diff
@@ -54,141 +55,331 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
 }
 
-void ExecutionHandler::stopImpl(StopReason&& stopReason)
+void ExecutionHandler::stopTheWorld(VM& targetVM, StopTheWorldEvent event)
```
What happens if two VMs (A, B) enter this function at the same time? Naively it seems like the following sequence is possible:

- A acquires the lock
- B fails to acquire the lock and goes to sleep
- A calls requestStopAll
- A releases the lock
- B wakes up (or would it be caught here and stopped? If so, I can still imagine a case where B comes into this function and only tries to acquire the lock right after A released it)
- B acquires the lock (this seems like a reasonable point to catch it, but I didn't see where that would happen)
- B calls requestStopAll -- what happens here? Could either event be lost?
- B releases the lock

hyjorc1 replied:
`VMManager::requestStopAll` just sets the stop request. And `VMManager::notifyVMStop` does the actual stop. So, 
```
A or B acquires/releases the lock doesn't matter
m_targetVM can be either A or B
A or B requestStopAll doesn't matter since CONCURRENT_SAFE
A and B both stop inside notifyVMStop
m_targetVM is the debuggee
```

hyjorc1 replied:
Key Safety Mechanisms:
- m_awaitingResumeNotification prevents overlapping stop-the-world operations
- requestStopAll() is idempotent - safe to call multiple times
- notifyVMStop() blocks until ALL VMs are stopped, preventing races

hyjorc1 replied:
Target VM Selection at Breakpoint:
The VM that last updated `ExecutionHandler::m_targetVM` is the debuggee. Since `notifyVMStop -> wasmDebuggerOnStopCallback -> handleStopTheWorld` always select the `ExecutionHandler::m_target` as the debuggee. See `selectTargetVMIfNeeded` for details.

kmiller68 commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -142,8 +144,13 @@ void initializeWithOptionsCustomization(const ScopedLambda<void()>& optionsCusto
                 Wasm::prepareSignalingMemory();
+#if ENABLE(WEBASSEMBLY)
+                VMManager::setWasmDebuggerOnStop(Wasm::wasmDebuggerOnStopCallback);
+                VMManager::setWasmDebuggerOnResume(Wasm::wasmDebuggerOnResumeCallback);
```
What's the point of having these in the config if they're only set here and always set when wasm is enabled?

hyjorc1 replied:
JSCConfig provides pointer authentication (via JSC_CONFIG_METHOD macro) for security and maintains architectural separation between `VMManager` and `WasmExecutionHandler`. This follows the pattern from the original [(stop-the-world implementation)](https://commits.webkit.org/301027@main) which put `memoryDebuggerStopTheWorld` in JSCConfig for the same reasons. So, while these callbacks are constant when `ENABLE(WEBASSEMBLY)` is true, the indirection provides security benefits and maintains architectural boundaries.

kmiller68 commented on Source/JavaScriptCore/wasm/debugger/WasmBreakpointManager.h:
```diff
@@ -32,28 +32,22 @@
-class IPIntCallee;
-class ModuleManager;
-
 class BreakpointManager {
```
Is anything in this class used outside JavaScriptCore.framework? Why are all the methods `JS_EXPORT_PRIVATE`?

Regardless if everything is exported you can just mark the whole class as `JS_EXPORT_PRIVATE` and all methods will be exported for you.

hyjorc1 replied:
Yes, it's used by the `testwasmdebugger` executable, which is a separate binary that links against JavaScriptCore.framework. Will mark the whole class as `JS_EXPORT_PRIVATE`.

kmiller68 commented on Source/JavaScriptCore/wasm/debugger/WasmDebugServerUtilities.h:
```diff
@@ -130,6 +131,98 @@ struct Breakpoint {
+    std::unique_ptr<const StopData> stopData { nullptr };
+
+    // Step-into tracking (for step debugging behavior)
+    StepIntoEvent stepIntoEvent { };
```
I don't think you need the `{ }` here since `StepIntoEvent` has a default constructor.

kmiller68 commented on Source/JavaScriptCore/wasm/debugger/WasmExecutionHandler.cpp:
```diff
@@ -54,141 +55,331 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-        m_vm.traps().cancelStop();
-        m_vm.setIsWasmStopWorldActive(false);
-    }
+using enum ExecutionHandler::DebuggerState;
```
I would avoid doing this since it might carry over to other .cpp files in the same unified source, which tends to cause issues when new files are added, etc

kmiller68 commented on Source/JavaScriptCore/wasm/debugger/WasmExecutionHandler.cpp:
```diff
@@ -197,110 +388,220 @@ void ExecutionHandler::interrupt()
+    if (!callerVM.debugState()->takeStepIntoCall())
+        return;
+
+    [&]() {
```
Why is this a lambda rather than a block?

hyjorc1 replied:
We may need to exit the lambda call early with `return`s.

hyjorc1 replied:
So the lambda avoids deep if-nesting (arrow anti-pattern) by enabling early returns for guard conditions. Both functions `setStepIntoBreakpointForCall` and `setStepIntoBreakpointForThrow` have 3+ guard clauses - without the lambda, the main logic would be nested 3+ levels deep, reducing readability.

kmiller68 commented on Source/JavaScriptCore/wasm/debugger/WasmExecutionHandler.cpp:
```diff
@@ -197,110 +388,220 @@ void ExecutionHandler::interrupt()
-        // The calls have two debugging behaviors:
-        // 2. Step-into: If the callee is a Wasm call.
+        // Regular calls have two debugging behaviors:
         // 1. Step-over: If the callee is a non-Wasm call.
```
This is the fallthrough right? Can we flip the order of these comments and/or clarify that in the comment.

kmiller68 commented on Source/JavaScriptCore/wasm/debugger/WasmExecutionHandler.cpp:
```diff
@@ -197,110 +388,220 @@ void ExecutionHandler::interrupt()
+    if (!throwVM.debugState()->takeStepIntoThrow())
+        return;
+
+    [&]() {
```
Ditto.

hyjorc1 replied:
We may need to exit the lambda call early with returns. This is to avoid deep if-nesting.

kmiller68 commented on Source/JavaScriptCore/wasm/debugger/WasmExecutionHandler.cpp:
```diff
@@ -197,110 +388,220 @@ void ExecutionHandler::interrupt()
+            return;
+        if (!throwVM.callFrameForCatch->callee().isNativeCallee())
+            return;
+        RefPtr wasmCallee = uncheckedDowncast<Wasm::Callee>(throwVM.callFrameForCatch->callee().asNativeCallee());
```
These could probably be a regular `downcast`, we don't care about the perf here and might as well make sure the system is in a sane state.

kmiller68 commented on Source/JavaScriptCore/wasm/debugger/WasmModuleManager.h:
```diff
@@ -50,31 +51,30 @@ class ModuleManager {
 public:
-    JS_EXPORT_PRIVATE ModuleManager(VM&);
-    JS_EXPORT_PRIVATE ~ModuleManager();
+    ModuleManager() = default;
```
Ditto on `JS_EXPORT_PRIVATE`

hyjorc1 replied:
Yes, it's used in `testwasmdebugger`. Will mark the whole class as `JS_EXPORT_PRIVATE`.

kmiller68 reviewed (approved):
r=me with comments

hyjorc1 committed:
[JSC][Wasm][Debugger] Implement multi-VM stop-the-world debugging
https://bugs.webkit.org/show_bug.cgi?id=302699
rdar://164945623

Reviewed by Keith Miller.

Implements comprehensive multi-VM stop-the-world Wasm debugger that stops ALL VMs
in the process when any VM hits a breakpoint or receives an interrupt.

Architecture:

* VMManager world modes coordinate multi-VM execution:
    - Running: Normal execution (all VMs running).
    - Stopped: All VMs paused at safe points for inspection.
    - RunOne: Single-step mode (target VM runs, others stopped).

* Two-phase callback system:
    - wasmDebuggerOnStop(): VMManager calls with ALL VMs stopped; returns resume mode.
    - wasmDebuggerOnResume(): VMManager calls after ALL VMs resume; releases barriers.

* Per-VM DebugState tracks execution state:
    - Running vs Stopped with stop reason (Prologue/Breakpoint/SystemCall).
    - StopData snapshot: PC, locals, stack, call frame for inspection.
    - Step-into event flags for cross-function stepping (call/throw).

* ExecutionHandler bridges LLDB commands to VM execution:
    - Detailed design in Debugger-Mutator-Protocol.md.
    - Two condition variables orchestrate debugger-mutator communication.
    - State machine (Replied/InterruptRequested/ContinueRequested/StepRequested/SwitchRequested).
    - Resume barrier prevents overlapping stop-the-world operations.
    - Target VM selection: prefers VM at prologue, falls back to triggering VM.

* Step-into design minimizes runtime changes:
    - step() sets event flag instead of resolving callee/handler immediately
    - Runtime resolves target naturally during execution (call dispatch, exception unwinding)
    - Callback sets breakpoint using resolved target, avoiding duplicate resolution logic
    - Reuses existing runtime mechanisms rather than duplicating complex resolution in debugger

Testing:

* Unit tests: ExecutionHandlerTest validates coordination logic.
* Integration tests: multi-vm test cases for same/different modules and functions.

Note that the WebAssembly debugger currently supports ARM64 platforms only.

Canonical link: https://commits.webkit.org/305160@main
